### PR TITLE
Implement find subprojects and do not recurse to a non-bare repository for src-expose serve

### PR DESCRIPTION
**Pull request by joshuabezaleel**
Fixes: #8636 

If a particular directory is a repository, proceed to configure and add to `gitDirs`.
If a particular repository is a bare repository, proceed to recurse. If it is a non-bare repository, skip it so that it won't recurse.

This is my first time contributing to Sourcegraph and I am still learning Go so I am really sorry beforehand if I got several things wrong. Looking forward to the review! :slightly_smiling_face:

### CODE REVIEW

File: dev/src-expose/serve.go [hunk 1]
```diff
@@ -138,29 +139,48 @@ func configureRepos(logger *log.Logger, root string) []string {
...
-		// stat now to avoid recursing into the rest of path
+
+		// Check whether a particular directory is a repository or not.
+		//
+		// A directory which also is a repository (have .git folder inside it)
+		// will contain nil error. If it does, proceed to configure.
 		gitdir := filepath.Join(path, ".git")
-		if _, err := os.Stat(gitdir); os.IsNotExist(err) {
-			return nil
-		}
```
Comment (unknwon) [inline dev/src-expose/serve.go]: We should preserve this error checking pattern.

File: dev/src-expose/serve.go [hunk 2]
```diff
@@ -138,29 +139,48 @@ func configureRepos(logger *log.Logger, root string) []string {
...
-		if err := configureOneRepo(logger, gitdir); err != nil {
-			logger.Printf("configuring repo at %s: %v", gitdir, err)
-			return nil
-		}
+		if _, err := os.Stat(gitdir); err == nil {
```
Comment (keegancsmith) [inline dev/src-expose/serve.go]: nit: by checking `err == nil` you have moved all the "main" logic of this function inside of this if statement body. This makes the code more indented, and generally distracts from the flow of the function. Could you keep the flow we previously had where we had early return

File: dev/src-expose/serve.go [hunk 3]
```diff
@@ -138,29 +139,48 @@ func configureRepos(logger *log.Logger, root string) []string {
...
+			// If it yields false, which means it is a non-bare repository,
+			// skip the directory so that it will not recurse to the subdirectories.
+			// If it is a bare repository, proceed to recurse.
+			c := exec.Command("git", "rev-parse", "--is-bare-repository")
+			c.Dir = gitdir
+			out, _ := c.CombinedOutput()
+
+			if string(out) == "false\n" {
+				return filepath.SkipDir
+			}
```
Comment (keegancsmith) [inline dev/src-expose/serve.go]: lgtm

**Comment by keegancsmith** [comment]
Just enabled the CI run, once green lets merge :) Thank you @joshuabezaleel!

**Comment by joshuabezaleel** [comment]
The thank you should be mine to say!! Thank you very much for all the helps and hospitality, @unknwon @keegancsmith !! :slightly_smiling_face: 
It's a really pleasant experience working on this and I would really love to be able to have more chance in the future contributing to Sourcegraph! :grin: 

My best regards to all of the Sourcegraph team :blush:

**Pull request merged**